File: LeetCode/1770. Maximum Score from Performing Multiplication Operations/solution.py
# class Solution:
#     def maximumScore(self, nums: List[int], multipliers: List[int]) -> int:
#         n, m = len(nums), len(multipliers)
#         @lru_cache(None)
#         def dfs(i,j):
#             if i == m:
#                 return 0
#             res = max(multipliers[i]*nums[j]+dfs(i+1,j+1),multipliers[i]*nums[j-i-1]+dfs(i+1,j))
#             # print(i,j,res)
#             return res
#         ans = dfs(0,0)
#         dfs.cache_clear()
#         return ans
    
    
class Solution:
    def maximumScore(self, nums: List[int], multipliers: List[int]) -> int:
        n, m = len(nums), len(multipliers)
        dp = [0]*(m+1)
        for i in range(m)[::-1]:
            for j in range(i+1):
                dp[j] = max(dp[j]+multipliers[i]*nums[j-i-1], dp[j+1]+multipliers[i]*nums[j])
        return dp[0]

File: LeetCode/1770. Maximum Score from Performing Multiplication Operations/solution.cpp
class Solution {
public:
    int maximumScore(vector<int>& nums, vector<int>& multipliers) {
        int n = nums.size(), m = multipliers.size();
        vector<int> dp(m+1, 0);
        for (int k = 1; k <= m; k++){
            dp[k] = dp[k-1] + nums[k-1]*multipliers[k-1];
            for (int i = k-1; i > 0; i--){
                dp[i] = max(dp[i-1]+nums[i-1]*multipliers[k-1],dp[i]+nums[n+i-k]*multipliers[k-1]);
            }
            dp[0] = dp[0] + nums[n-k]*multipliers[k-1];
        }
        return *max_element(dp.begin(), dp.end());
    }
};
